### PR TITLE
feat(pipeline): allow additional policy to build role

### LINKED ISSUE #2755: Feature request: build project role addon policies
It is well documented that you can create an addon with a ManagedPolicy resource and an Output definition and this policy will be injected into the ECS Task Role. As documented on this page: https://aws.github.io/copilot-cli/docs/developing/additional-aws-resources/#what-does-an-addon-template-look-like

> If you need to add additional policies to your ECS task role, you can define an IAM ManagedPolicy resource in your template that holds the permissions for your task, and then add it as an Output. The permission will be injected into your ECS Task Role.

I am working with pipelines and wish to add permissions to the CodeBuild Build Project Role so I can grant, for example, cross-account CodeArtifact access and allow the Build Project Role to perform a login similar to:

`aws codeartifact login --tool pip --repository artifacts --domain <domain> --domain-owner <owner>`

I asked about this in the the project discussion: https://github.com/aws/copilot-cli/discussions/2535

It seems this is not something that is currently supported.

Proposal:
Offer a way to configure Outputs for addon templates that can attach to the Build Project Role and continue to support Task Role injection as well.

### PR DESCRIPTION
**Pull request by paragbhingre**
Resolves #2755

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/templates/cicd/pipeline_cfn.yml [hunk 1]
```diff
@@ -124,6 +124,22 @@ Resources:
           {{- end }} {{/* endif ne .Source.OutputArtifactFormat "" */}}{{- end }} {{/* endif ne .Source.ProviderName "GitHubV1" */}}
       Roles:
         - !Ref BuildProjectRole
+  {{- if isSet .Build.AdditionalPolicy }}
```
Comment (dannyrandall) [inline internal/pkg/template/templates/cicd/pipeline_cfn.yml]:
```suggestion
  {{- if .Build.AdditionalPolicy }}
```
i don't think you need `isSet`

File: internal/pkg/manifest/pipeline.go [hunk 2]
```diff
@@ -178,8 +183,9 @@ type Source struct {
 
 // Build defines the build project to build and test image.
 type Build struct {
-	Image     string `yaml:"image"`
-	Buildspec string `yaml:"buildspec,omitempty"`
+	Image            string            `yaml:"image"`
+	Buildspec        string            `yaml:"buildspec,omitempty"`
+	AdditionalPolicy *AdditionalPolicy `yaml:"additional_policy,omitempty"`
```
Comment (dannyrandall) [inline internal/pkg/manifest/pipeline.go]: what would you think about just having this as `yaml.Node` or `map[string]any` and marshaling it directly into the pipeline template? and just letting cloudformation do the validation?

Comment (iamhopaul123) [reply internal/pkg/manifest/pipeline.go]: Agreed. I think right now it is pretty much the same as what CFN has (we don't do any abstraction or setting default values for them). We could just let them specify [the json policy document](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-policy.html#cfn-iam-policy-policydocument) as a string.

Comment (paragbhingre) [reply internal/pkg/manifest/pipeline.go]: The only thing I thought of is CFN template errors are little difficult for the customers to understand and solve that is why I kept it this way that we will guide customers with more user friendly errors.

Comment (iamhopaul123) [reply internal/pkg/manifest/pipeline.go]: Yeah but additional policies is a pretty advanced feature and most likely those users who need to specify additional policies they already have the json policy document handy.

File: internal/pkg/deploy/pipeline_test.go [hunk 3]
```diff
@@ -241,6 +241,18 @@ func TestPipelineBuild_Init(t *testing.T) {
 				BuildspecPath:   "some/path",
 			},
 		},
+		"additional policy is empty": {
```
Comment (efekarakus) [inline internal/pkg/deploy/pipeline_test.go:271]: should we have a test case where we are marshaling the `mfBuild.AdditionalPolicy`

File: internal/pkg/deploy/pipeline.go [hunk 4]
```diff
@@ -99,9 +101,18 @@ func (b *Build) Init(mfBuild *manifest.Build, mfDirPath string) {
 	if strings.Contains(image, "aarch64") {
 		environmentType = "ARM_CONTAINER"
 	}
+	if mfBuild != nil && mfBuild.AdditionalPolicy.Content != nil {
```
Comment (efekarakus) [inline internal/pkg/deploy/pipeline.go]:
```suggestion
	if mfBuild != nil && !mfBuild.AdditionalPolicy.IsZero() {
```

File: internal/pkg/template/templates/cicd/pipeline_cfn.yml [hunk 5]
```diff
@@ -124,6 +124,14 @@ Resources:
           {{- end }} {{/* endif ne .Source.OutputArtifactFormat "" */}}{{- end }} {{/* endif ne .Source.ProviderName "GitHubV1" */}}
       Roles:
         - !Ref BuildProjectRole
+  {{- if .Build.AdditionalPolicy }}
+  BuildProjectAdditionalPolicy:
+    Type: 'AWS::IAM::Policy'
+    Properties:
+      PolicyName: !Sub ${AWS::StackName}-BuildProjectAdditionalPolicy
+  {{indent 4 .Build.AdditionalPolicy}}
```
Comment (efekarakus) [inline internal/pkg/template/templates/cicd/pipeline_cfn.yml]: If I'm understanding correctly, this is assume that clients will write the pipeline manifest like this:
```yaml
build:
  additional_policy:
    PolicyDocument:
      Statement: ...
```

It's not super clear why `Roles` and `PolicyName` cannot be set in `additional_policy`. Can we instead change the struct to be something like this?
```go
type Build struct {
   // ..
   AdditionalPolicy struct {
      Document yaml.Node `yaml:"document",omitempty`
   } `yaml:"additional_policy,omitempty"`
}
```
The above makes it clear to me that only `PolicyDocument` will be customizable by clients

**Review by efekarakus — COMMENTED** [review]
Yay awesome, this will provide a lot more flexibility to customers! 🎉 

Is it also possible to add a local integration test in for example https://github.com/aws/copilot-cli/blob/1caa364fd59f1bab4e0bc6a09d1d271fff24f5ab/internal/pkg/deploy/cloudformation/stack/bb_pipeline_integration_test.go#L25 to make sure the additonal policy is rendered in the template?

File: internal/pkg/template/templates/cicd/pipeline_cfn.yml (hunk 5 above)
Comment (paragbhingre) [reply internal/pkg/template/templates/cicd/pipeline_cfn.yml]: Yeah this looks clean, I will change it.

File: internal/pkg/deploy/pipeline.go [hunk 6]
```diff
@@ -99,9 +101,18 @@ func (b *Build) Init(mfBuild *manifest.Build, mfDirPath string) {
 	if strings.Contains(image, "aarch64") {
 		environmentType = "ARM_CONTAINER"
 	}
+	if mfBuild != nil && !mfBuild.AdditionalPolicy.Document.IsZero() {
+		additionalPolicy, err := yaml.Marshal(&mfBuild.AdditionalPolicy.Document)
+		if err != nil {
+			return fmt.Errorf("marshal pipeline manifest to embed in template: %v", err)
```
Comment (efekarakus) [inline internal/pkg/deploy/pipeline.go]:
```suggestion
			return fmt.Errorf("marshal `additional_policy.PolicyDocument` in pipeline manifest: %v", err)
```

File: internal/pkg/template/templates/cicd/pipeline_cfn.yml [hunk 7]
```diff
@@ -124,6 +124,15 @@ Resources:
...
         - !Ref BuildProjectRole
+  {{- if .Build.AdditionalPolicyDocument }}
+  BuildProjectAdditionalPolicy:
+    Type: 'AWS::IAM::Policy'
+    Properties:
+      PolicyName: !Sub ${AWS::StackName}-BuildProjectAdditionalPolicy
+      PolicyDocument:
+{{.Build.AdditionalPolicyDocument | indent 8}}
+      Roles:
+        - !Ref BuildProjectRole {{- end}}
```
Comment (efekarakus) [inline internal/pkg/template/templates/cicd/pipeline_cfn.yml]: nit:
```suggestion
        - !Ref BuildProjectRole
  {{- end}}
```

**Review by efekarakus — COMMENTED** [review]
Looks great! We can :shipit: after fixing the integ test

File: internal/pkg/template/templates/cicd/pipeline_cfn.yml [hunk 8]
```diff
@@ -124,6 +124,16 @@ Resources:
...
+  {{- if .Build.AdditionalPolicyDocument }}
+  BuildProjectAdditionalPolicy:
+    Type: 'AWS::IAM::Policy'
+    Properties:
+      PolicyName: !Sub ${AWS::StackName}-BuildProjectAdditionalPolicy
+      PolicyDocument:
+{{.Build.AdditionalPolicyDocument | indent 8}}
+      Roles:
+        - !Ref BuildProjectRole
+{{- end}}
```
Comment (efekarakus) [inline internal/pkg/template/templates/cicd/pipeline_cfn.yml]:
```suggestion
  {{- end}}
```

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/deploy/pipeline.go [hunk 9]
```diff
@@ -99,9 +101,18 @@ func (b *Build) Init(mfBuild *manifest.Build, mfDirPath string) {
 	if strings.Contains(image, "aarch64") {
 		environmentType = "ARM_CONTAINER"
 	}
+	if mfBuild != nil && !mfBuild.AdditionalPolicy.Document.IsZero() {
+		additionalPolicy, err := yaml.Marshal(&mfBuild.AdditionalPolicy.Document)
```
Comment (dannyrandall) [inline internal/pkg/deploy/pipeline.go:105]: do we need to pass the pointer to the document? could we just do `mfBuild.AdditionalPolicy.Document`?

File: internal/pkg/deploy/pipeline_test.go [hunk 10]
```diff
@@ -203,6 +204,17 @@ func TestPipelineBuild_Init(t *testing.T) {
...
+	yamlNode := yaml.Node{}
+	policyDocument := []byte(`
+  Statement:
+    Action: '*'
+    Effect: Allow
+    Resource: '*'
+  Version: 2012-10-17`)
+
+	if err := yaml.Unmarshal(policyDocument, &yamlNode); err != nil {
+		t.Errorf("unmarshal error:  %v", err)
```
Comment (dannyrandall) [inline internal/pkg/deploy/pipeline_test.go]: let's do `require.NoError(t, err)`, just to be consistent with the other failures

File: internal/pkg/deploy/cloudformation/stack/dd_pipeline_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+				KeyArn:     "arn:aws:kms:us-west-2:1111:key/abcd",
+			},
+		},
+		AdditionalTags: nil,
+	})
+
+	actual, err := serializer.Template()
+	require.NoError(t, err, "template should have rendered successfully")
+	actualInBytes := []byte(actual)
+	m1 := make(map[interface{}]interface{})
```
Comment (dannyrandall) [inline internal/pkg/deploy/cloudformation/stack/dd_pipeline_integration_test.go]:
```suggestion
	m1 := make(map[any]any)
```

**Review by dannyrandall — APPROVED** [review]
just some small things but looks good!!

**Comment by huanjani** [comment]
Resolves https://github.com/aws/copilot-cli/issues/1689, https://github.com/aws/copilot-cli/issues/2546, https://github.com/aws/copilot-cli/discussions/2535, https://github.com/aws/copilot-cli/discussions/2104

File: internal/pkg/deploy/pipeline_test.go (hunk 10 above)
Comment (paragbhingre) [reply internal/pkg/deploy/pipeline_test.go]: I think we need `require.Error(t, err)` as we are going inside if only when error occurs. What do you think about it?

File: internal/pkg/deploy/pipeline.go (hunk 9 above)
Comment (paragbhingre) [reply internal/pkg/deploy/pipeline.go:105]: Actually when we `unmarshal` the whole manifest we point to the `Pipeline` variable in which we store the unmarshalled manifest. And while we `Marshal` is back into bytes then also we need pointer to the same variable. So indeed we need `&mfBuild.AdditionalPolicy.Document` else we get empty response from `Marshal`. 

Does that make sense?

File: internal/pkg/deploy/pipeline_test.go (hunk 10 above)
Comment (dannyrandall) [reply internal/pkg/deploy/pipeline_test.go]: i think you can remove the if check and just do something like this:
```
require.NoError(t, yaml.Unmarshal(policyDocument, &yamlNode))
```

**Pull request merged**